### PR TITLE
Creates previously used keywords plugin

### LINKED ISSUE #370: Create assessment for keyword doubles
Create an assessment for keyword doubles

This assessment doesn't really belong  in the YoastSEO.js since it doesn't assess the paper, but uses a wordpress specific array with keywords to see if they are use before.
It might be better to include this in wordpress-seo as an extra wordpress specific assessment.

### PR DESCRIPTION
**Pull request by terw-dan**
Creates a plugin for previously used keywords that can be registered to pluggable and the assessment added to the assessor. 

Fixes #370

### CODE REVIEW

File: js/app.js [hunk 1]
```diff
@@ -594,16 +594,29 @@ App.prototype.registerModification = function( modification, callable, pluginNam
...
- * @param {Object}   scoring    A scoring object that defines how the analysis translates to a certain SEO score.
+ *
+ * @deprecated
+ */
+App.prototype.registerTest = function() {
+	console.error( "This function is deprecated, please use registerAssessment" );
+};
+
+/**
+ * Registers a custom test for use in the analyzer, this will result in a new line in the analyzer results. The function
```
Comment (atimmer) [inline js/app.js]: Please rephrase this to be more clear that you can add a custom assessment, not a test or something to the analyser.

File: js/app.js [hunk 2]
```diff
@@ -594,16 +594,29 @@ App.prototype.registerModification = function( modification, callable, pluginNam
...
+ * has to return a result based on the contents of the page/posts.
+ *
+ * Score 0 results in a grey circle
```
Comment (atimmer) [inline js/app.js]: Score 0 only results in a grey circle if it has not been explicitly set using `setScore`, otherwise the circle is red.

File: js/app.js [hunk 3]
```diff
@@ -594,16 +594,29 @@ App.prototype.registerModification = function( modification, callable, pluginNam
...
+ * Scores 1, 2, 3 and 4 result in a red circle
+ * Scores 6 and 7 result in a yellow circle
+ * Scores 8, 9 and 10 result in a green circle
```
Comment (atimmer) [inline js/app.js]: Please document that the assessment should return an AssessmentResult.

File: js/assessor.js [hunk 4]
```diff
@@ -152,4 +152,17 @@ Assessor.prototype.calculateOverallScore  = function() {
 	return Math.round( totalScore / ( results.length * ScoreRating ) * 100 );
 };
 
+/**
+ * Register an assessment for a specific plugin
```
Comment (atimmer) [inline js/assessor.js]: The assessor knows nothing about plugins, it will just add an assessment to its internal assessment object.

File: js/bundledPlugins/previouslyUsedKeywords.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
+var AssessmentResult = require( "../values/AssessmentResult.js" );
+var isUndefined = require( "lodash/isUndefined" );
+
+/**
+ * Constructs the PreviouslyUsedKeywords plugin
```
Comment (atimmer) [inline js/bundledPlugins/previouslyUsedKeywords.js]: This documentation is redundant. We should document what you can do with a PreviouslyUsedKeywords plugin.

File: js/bundledPlugins/previouslyUsedKeywords.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+ * @param {object} app The app
+ * @param {object} usedKeywords An object with keywords and ids where they are used
+ * @param {string} searchUrl The url used to link to a search page when multiple usages of the keyword are found
+ * @param {string} postUrl The url used to link to a post when 1 usage of the keyword is found
+ * @constructor
+ */
+var PreviouslyUsedKeywords = function( app, usedKeywords, searchUrl, postUrl ) {
```
Comment (atimmer) [inline js/bundledPlugins/previouslyUsedKeywords.js]: Please drop the s.

File: js/bundledPlugins/previouslyUsedKeywords.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+	this.app = app;
+	this.usedKeywords = usedKeywords;
+	this.searchUrl = searchUrl;
+	this.postUrl = postUrl;
+	this.i18n = this.app.i18n;
```
Comment (atimmer) [inline js/bundledPlugins/previouslyUsedKeywords.js]: Why does this object need to know where the i18n object is located on the app. i18n should just be a parameter.

File: js/bundledPlugins/previouslyUsedKeywords.js (hunk 6 above)
Comment (atimmer) [inline js/bundledPlugins/previouslyUsedKeywords.js]: Please combine `usedKeywords`, `searchUrl` and `postUrl` into one argument.

File: js/bundledPlugins/previouslyUsedKeywords.js [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+		id = this.usedKeywords[ keyword ][ 0 ];
+	}
+	var result = this.scoreAssessment( count, id, keyword );
+
+	var assessmentResult =  new AssessmentResult();
+	assessmentResult.setScore( result.score );
+	assessmentResult.setText( result.text );
+
+	return assessmentResult;
+};
```
Comment (atimmer) [inline js/bundledPlugins/previouslyUsedKeywords.js]: We should add a function to this prototype, `addKeywordUsage` to have an API to add keyword usage information.

File: js/pluggable.js [hunk 9]
```diff
@@ -170,48 +173,40 @@ Pluggable.prototype._registerModification = function( modification, callable, pl
...
+ * @deprecated
+ */
+Pluggable.prototype._registerTest = function() {
+	console.error ( "This function is deprecated, please use _registerAssessment" );
+};
+
+/**
+ * Register an assessment for a specific plugin
+ *
+ * @param {string} name The name of the assessment.
```
Comment (atimmer) [inline js/pluggable.js]: Please document the `assessor` parameter.

File: spec/pluggableSpec.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+		beforeEach( function() {
+			app = {
+				updateLoadingDialog: function() {}
+			};
+			pluggable = new Pluggable( app );
+			pluggable._registerPlugin( "test-plugin" );
+		});
+
+		it( "should throw an error on adding an invalid assessment", function() {
+			expect( function() { pluggable._registerAssessment( false ) } ).toThrowError( invalidTypeError );
```
Comment (atimmer) [inline spec/pluggableSpec.js]: Please capitalise `invalidTypeError` so it's clear that it is a prototype.

File: js/bundledPlugins/previouslyUsedKeywords.js [hunk 11]
```diff
@@ -0,0 +1,92 @@
+var AssessmentResult = require( "../values/AssessmentResult.js" );
+var isUndefined = require( "lodash/isUndefined" );
+
+/**
+ * @param {object} app The app
+ * @param {object} args An arguments object with usedKeywords, searchUrl, postUrl,
```
Comment (jcomack) [inline js/bundledPlugins/previouslyUsedKeywords.js]: Please add separate param options for the possible properties within the args object.

See scoreformatter for an example of this.

File: js/bundledPlugins/previouslyUsedKeywords.js [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+/**
+ * Registers the assessment with the assessor.
+ */
+PreviouslyUsedKeyword.prototype.registerPlugin = function() {
+	this.app.registerAssessment( "usedKeywords", this.assess, "previouslyUsedKeywords" );
+};
+
+/**
+ * Replaces the usedKeywords
```
Comment (jcomack) [inline js/bundledPlugins/previouslyUsedKeywords.js]: Please change this to "Updates the usedKeywords"

File: js/bundledPlugins/previouslyUsedKeywords.js [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+ */
+PreviouslyUsedKeyword.prototype.scoreAssessment = function( count, id, keyword ) {
+	if( count === 0 ) {
+		return {
+			text: this.i18n.dgettext( "js-text-analysis", "You've never used this focus keyword before, very good." ),
+			score: 9
+		};
+	}
+	if( count === 1 ) {
+		var url = "<a href='" + this.postUrl.replace( "{id}", id ) + "'>";
```
Comment (jcomack) [inline js/bundledPlugins/previouslyUsedKeywords.js]: Possibly extract this replacement action and https://github.com/Yoast/YoastSEO.js/pull/395/files#diff-9da8cedbf92f047bb549025aca5957aeR57. This would also make the defining of the url variable redundant and would lead to cleaner, more readable code.

File: js/bundledPlugins/previouslyUsedKeywords.js [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+
+/**
+ * The assessment for the previously used keywords.
+ * @returns {AssessmentResult} The assessment result of the assessment
+ */
+PreviouslyUsedKeyword.prototype.assess = function() {
+	var keyword = this.app.paper.getKeyword();
+	var count = 0;
+	var id = 0;
+	if ( !isUndefined( this.usedKeywords[ keyword ] ) ) {
```
Comment (jcomack) [inline js/bundledPlugins/previouslyUsedKeywords.js]: These if statements could quite possibly be moved into the `scoreAssessment` function. This logic doesn't seem fitting to be here.

**Comment by jcomack** [comment]
CR done

**Comment by jcomack** [comment]
CR 2 done :+1:

**Comment by jcomack** [comment]
Couldn't complete Acceptance due to the following:
- Code is never included in the example file, thus cannot be tested
- When manually adding the file, a config is necessary
- When manually adding the file, JS throws an error `Uncaught ReferenceError: require is not defined`. This is probably due to browserify not properly including the bundledPlugins folder

**Comment by terw-dan** [comment]
For acceptance, please use the browserified example

File: examples/browserified/example.js [hunk 15]
```diff
@@ -21,5 +32,18 @@ window.onload = function() {
 		}
 	});
 
+	bindEvents( app );
+
 	app.refresh();
+
+	var args ={
```
Comment (andizer) [inline examples/browserified/example.js]: Please add a space before `{`

File: js/app.js [hunk 16]
```diff
@@ -568,16 +568,29 @@ App.prototype.registerModification = function( modification, callable, pluginNam
  * Scores 6 and 7 result in a yellow circle
  * Scores 8, 9 and 10 result in a red circle
  *
- * @param {string}   name       Name of the test.
- * @param {function} analysis   A function that analyzes the content and determines a score for a certain trait.
- * @param {Object}   scoring    A scoring object that defines how the analysis translates to a certain SEO score.
+ *
+ * @deprecated
```
Comment (andizer) [inline js/app.js]: Can you add a version number to clear at which version this is deprecated.

File: examples/browserified/example.js [hunk 17]
```diff
@@ -1,5 +1,16 @@
 var SnippetPreview = require( "../../js/snippetPreview" );
 var App = require( "../../js/app" );
+var PreviouslyUsedKeywords = require( "../../js/bundledPlugins/previouslyUsedKeywords.js" );
+
+/**
+ * binds the renewData function on the change of inputelements.
```
Comment (andizer) [inline examples/browserified/example.js]: I do not understand what you mean. I guess you are binding it to an event?

File: js/bundledPlugins/previouslyUsedKeywords.js [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+ * @param {string} args.postUrl The url used to link to a post when 1 usage of the keyword is found.
+ * @param {Assessor} args.assessor
+ * @param {object} i18n The i18n object used for translations
+ * @constructor
+ */
+var PreviouslyUsedKeyword = function( app, args, i18n ) {
+	if ( isUndefined( app ) ) {
+		throw new MissingArgument( "The previously keyword plugin requires the YoastSEO app" );
+	}
+	if ( isUndefined( args ) ) {
```
Comment (andizer) [inline js/bundledPlugins/previouslyUsedKeywords.js]: I would move this below the `i18n` if. Because the `app` and `i18n` are required.

File: js/pluggable.js [hunk 19]
```diff
@@ -170,48 +173,41 @@ Pluggable.prototype._registerModification = function( modification, callable, pl
...
+		throw new InvalidTypeError( "Failed to register assessment for plugin " + pluginName +
+			". Expected parameter `assessment` to be a function." );
 	}
 
-	if ( typeof pluginName !== "string" ) {
-		console.error( "Failed to register test for plugin " + pluginName + ". Expected parameter `pluginName` to be a string." );
-		return false;
+	if ( !isString( pluginName ) ) {
+		throw new InvalidTypeError( "Failed to register assessment for plugin " + pluginName +
+			". Expected parameter `pluginName` to be a strign." );
```
Comment (andizer) [inline js/pluggable.js]: `strign` ?

**Comment by andizer** [comment]
CR done
